### DISCUSSION TITLE
DB connections and pooling

### DISCUSSION THREAD

**Discussion by ghelle-fhh** [ldproxy/ldproxy#744] [Configuration]
Hi,
I've been testing out ldproxy recently and really appreciate all the good work you've done so far.

However, I'm running into an issue with the DB connections. 

We want to publish 200-300 datasets stored in a PostGIS Database via ldproxy - each dataset is configured as a provider. The datasets are typically stored in a seperated schema.

I'm running into the problem that 1 provider per dataset leads to a very large number of DB connections (200-300), even though the datasets share the same host, pw, and database. 

This problem becomes especially acute when running  Ldproxy in Kubernetes with multiple pods, i.e. instances, as it can lead to potentially thousands of DB connections, leading to a bottleneck on the server running the database.

Reading the docs and in practical tests it seems I can only pool the DB connections if the connection info is exactly the same - and if the schema is different start-up of other providers fails.

Is there already a better way to pool the providers connecting with the database, or am I missing something?

Thanks!

**Comment by azahnen** [comment]
Hi,
that is correct, connection pools can only be shared between providers when connection info is exactly the same.
If your table names were unique across schemas, which I guess they are not, you could just use the same schema list in all providers. 
Otherwise you would need schema prefixes for table mappings, which are not supported yet. That just came up in another discussion #742, where I proposed multiple providers as a workaround.

**Comment by ghelle-fhh** [comment]
Hello,

Thank you again for your response.

Indeed our table names are not unique across schemas, so schema prefixes would be required. 

However we also have the additional problem that our data is spread across multiple (3-4) databases. So even if schema prefixes were enabled the connection info would be different and fail for any additional databases.

In this case we would need multiple shared connection pools for multiple connections, which already seems to be opened as an enhancement (#732), albeit the story does not seem to match our use-case.

**Comment by azahnen** [comment]
FYI, [v3.3.0](https://github.com/interactive-instruments/ldproxy/releases/tag/v3.3.0) contains [schema prefixes](https://github.com/interactive-instruments/ldproxy/issues/772) and [entity groups](https://github.com/interactive-instruments/ldproxy/issues/773), which should provide a clean way to implement your use case. Let me know when you need any help.

**Comment by bgrotan** [comment]
Have you considered an external connection-pool such as pgBouncer or pgPool, to offload the database servers?
https://www.linkedin.com/pulse/postgresql-connection-pooling-part-4-pgbouncer-vs-kristi-anderson/

**Reply by ghelle-fhh** [reply]
Thanks for the tip! 

In fact we have. We did some intial tests with it and it appeared to perform fine and solve the issue. We just were unsure if we really want to operate an additional component to solve this issue.
